Define new TYPE and UPLINK vars for registration

nkinkade opened the pull request:
This change allows Autojoin clients to specify the machine type and the uplink speed. This information will be passed to Locate via Heartbeat messages.

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/m-lab/autonode/27)
<!-- Reviewable:end -->

nkinkade committed:
Defines new TYPE and UPLINK vars for registration

The Autojoin API's register endpoint is going to start enforcing the presence
of these values. Today, all M-Lab managed machines has these values as part of
the Heartbeat registration process, and so should Autojoin machines.

nkinkade committed:
Adds TYPE and UPLINK vars to .env file for deploy.sh

nkinkade committed:
Temporarily sets DOCKER_TAG=latest for testing

nkinkade committed:
Revert "Temporarily sets DOCKER_TAG=latest for testing"

This reverts commit 032ab1cbfc41c20bb1ed239ec2d236cbf28641d4.

nkinkade reviewed (commented):
FWIW: I temporarily set DOCKER_TAG=latest and pushed. The build succeeded, and the output from Locate:/v2/siteinfo/registrations for the mlab test node in sandbox did have those two fields populated as expected. 

> *[Reviewable](https://reviewable.io/reviews/m-lab/autonode/27#-:-OCkSlyNBQ_FwHjCmsjS:b-tbssvw)* status: 0 of 1 approvals obtained (waiting on @stephen-soltesz)



<!-- Sent from Reviewable.io -->

stephen-soltesz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/m-lab/autonode/27)* status: 0 of 1 approvals obtained (waiting on @nkinkade)

___
*[`deploy.sh` line 65 at r1](https://reviewable.io/reviews/m-lab/autonode/27#-OCkSCicB8iPgciWuAEp:-OCkSCicB8iPgciWuAEq:bs62nd9) ([raw file](https://github.com/m-lab/autonode/blob/9c0f986271577c5f1f43e707a107e524ac2f1cb9/deploy.sh#L65)):*
> ```Shell
>     echo "IPV6=\$IPV6" >> .env
>     echo "TYPE=virtual" >> .env
>     echo "UPLINK=1g" >> .env
> ```

Do we use 1g for ndt-virtual VMs?
___
*[`examples/ndt-fullstack.yml` line 29 at r1](https://reviewable.io/reviews/m-lab/autonode/27#-OCkSIHk7bkxuTdfUusC:-OCkSIHk7bkxuTdfUusD:b3kn4pa) ([raw file](https://github.com/m-lab/autonode/blob/9c0f986271577c5f1f43e707a107e524ac2f1cb9/examples/ndt-fullstack.yml#L29)):*
> ```YAML
>       - -ipv6=${IPV6}
>       - -type=${TYPE}
>       - -uplink=${UPLINK}
> ```

Hmm, this is tricky. 

Right now the `DOCKER_TAG=stable` which won't have this flag yet... This is why the initial deployment of the equinix site didn't have service account keys - it was using an old version of register (the previous "stable") which didn't write the service account yet.

I'm considering that it may be preferable to do away with 'stable' since it maybe a source of more confusion than help.

I'm inclined to recommend we replace "stable" with the latest tagged version of register that supports these flags.

The original intention behind "stable" is that it would provide a remote update path for this functionality and it would decouple our releases of autonode from the already deployed versions in the world. But, it requires a manual step that we forgot once, and maybe this is an example of a second time (just b/c you probably didn't know about this).

In principle, deploy.sh could find the latest tag using something like:

```
$ curl -s 'https://api.github.com/repos/m-lab/autonode/releases?per_page=1' | jq '.[].name'
```

What do you think?


<!-- Sent from Reviewable.io -->

nkinkade reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/m-lab/autonode/27)* status: 0 of 1 approvals obtained (waiting on @stephen-soltesz)

___
*[`deploy.sh` line 65 at r1](https://reviewable.io/reviews/m-lab/autonode/27#-OCkSCicB8iPgciWuAEp:-OCkVVgZ-7Q_MICoRitM:b-iuda0t) ([raw file](https://github.com/m-lab/autonode/blob/9c0f986271577c5f1f43e707a107e524ac2f1cb9/deploy.sh#L65)):*
<details><summary><i>Previously, stephen-soltesz (Stephen Soltesz) wrote…</i></summary><blockquote>

Do we use 1g for ndt-virtual VMs?
</blockquote></details>

It isn't accurate, but this is what we have been doing for GCP VMs for the past couple years. I can't recall at the moment the rationale for doing this. For example:

https://github.com/m-lab/siteinfo/blob/main/sites/pdx03.jsonnet#L23

I was just trying to replicate what we do today for GCP VMs, but I'm completely open to revisit why we're doing this.
___
*[`examples/ndt-fullstack.yml` line 29 at r1](https://reviewable.io/reviews/m-lab/autonode/27#-OCkSIHk7bkxuTdfUusC:-OCkW1Er-TYL1uTFeOEF:bhr2p9a) ([raw file](https://github.com/m-lab/autonode/blob/9c0f986271577c5f1f43e707a107e524ac2f1cb9/examples/ndt-fullstack.yml#L29)):*
<details><summary><i>Previously, stephen-soltesz (Stephen Soltesz) wrote…</i></summary><blockquote>

Hmm, this is tricky. 

Right now the `DOCKER_TAG=stable` which won't have this flag yet... This is why the initial deployment of the equinix site didn't have service account keys - it was using an old version of register (the previous "stable") which didn't write the service account yet.

I'm considering that it may be preferable to do away with 'stable' since it maybe a source of more confusion than help.

I'm inclined to recommend we replace "stable" with the latest tagged version of register that supports these flags.

The original intention behind "stable" is that it would provide a remote update path for this functionality and it would decouple our releases of autonode from the already deployed versions in the world. But, it requires a manual step that we forgot once, and maybe this is an example of a second time (just b/c you probably didn't know about this).

In principle, deploy.sh could find the latest tag using something like:

```
$ curl -s 'https://api.github.com/repos/m-lab/autonode/releases?per_page=1' | jq '.[].name'
```

What do you think?
</blockquote></details>

From what I can tell, the Docker Hub container image with tag "stable" is identical to the tag of the latest autojoin repo release. I don't think it would make any difference to reference the tag name instead of "stable". Am I looking at this incorrectly? 

In your example above you are referencing the "autonode" repo, but I think you meant to specify the "autojoin" repo, no?

What if deploy.sh could switch on PROJECT, where in mlab-sandbox and mlab-staging it would simply use the "latest" tag, and in production it would refer to "stable"?


<!-- Sent from Reviewable.io -->

stephen-soltesz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/m-lab/autonode/27)* status: 0 of 1 approvals obtained (waiting on @nkinkade)

___
*[`deploy.sh` line 65 at r1](https://reviewable.io/reviews/m-lab/autonode/27#-OCkSCicB8iPgciWuAEp:-ODC634J2R4sfMAekft7:bbmyu5v) ([raw file](https://github.com/m-lab/autonode/blob/9c0f986271577c5f1f43e707a107e524ac2f1cb9/deploy.sh#L65)):*
<details><summary><i>Previously, nkinkade wrote…</i></summary><blockquote>

It isn't accurate, but this is what we have been doing for GCP VMs for the past couple years. I can't recall at the moment the rationale for doing this. For example:

https://github.com/m-lab/siteinfo/blob/main/sites/pdx03.jsonnet#L23

I was just trying to replicate what we do today for GCP VMs, but I'm completely open to revisit why we're doing this.
</blockquote></details>

It just seems strange for there to be a disconnect between uplink and interface maxrate.

I seem to recall at the very beginning of including the VMs I may have advocated for starting with 1g (including the max rate) so we wouldn't direct too much traffic there to start. We increased the probability and max rate for vms but never revisited this annotation.

IIRC, our VMs have a max egress rate of 7g? Is that correct? Could the "UPLINK" value match the max egress rates for our various configurations? https://cloud.google.com/compute/docs/network-bandwidth#vm-out 

I guess "UPLINK" is a remnant of switch deployments. Today this is something more generic like, "EGRESS" or MAXEGRESS or similar.
___
*[`examples/ndt-fullstack.yml` line 29 at r1](https://reviewable.io/reviews/m-lab/autonode/27#-OCkSIHk7bkxuTdfUusC:-ODC7j5V4Ukt12xPVd4m:b-oz9xnk) ([raw file](https://github.com/m-lab/autonode/blob/9c0f986271577c5f1f43e707a107e524ac2f1cb9/examples/ndt-fullstack.yml#L29)):*
<details><summary><i>Previously, nkinkade wrote…</i></summary><blockquote>

From what I can tell, the Docker Hub container image with tag "stable" is identical to the tag of the latest autojoin repo release. I don't think it would make any difference to reference the tag name instead of "stable". Am I looking at this incorrectly? 

In your example above you are referencing the "autonode" repo, but I think you meant to specify the "autojoin" repo, no?

What if deploy.sh could switch on PROJECT, where in mlab-sandbox and mlab-staging it would simply use the "latest" tag, and in production it would refer to "stable"?
</blockquote></details>

I think the `stable` tag was a mistake. I would suggest we remove it and not include any remote update path to avoid the operational overhead of tracking a separate deployment path, and this would ensure that deployed configurations are stable. We would only need to maintain API compatibility in the Autojoin API.

I meant autonode because (in this model) M-Lab wants to be able to deploy a new version of the autonode to test it before promoting that to "stable" for all deployments. But, I think this is a bad idea. I think it would be preferable for there to just be the latest static version tag that autonode uses. Or, if deployed versions don't use the latest, it just continues to use the older version.


<!-- Sent from Reviewable.io -->

nkinkade committed:
Removes manual DOCKER_TAG var from cloudbuild.conf and deploy.sh

The value of this variable will now be determined dymaically at build time by
querying the repository directly for the latest tag.

nkinkade committed:
Makes DOCKER_TAG a user-editable var in Docker env file

We want to no longer user the "stable" tag. This prompts users to manually
verify the latest release of the autojoin repository and use that as the value
of the variable DOCKER_TAG.

nkinkade committed:
Set UPLINK=7g in deploy.sh

This value is more consistent with the actual max uplink rate for GCP VMs.
Also, updates USAGE, removing "<docker-tag>".

nkinkade reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/m-lab/autonode/27)* status: 0 of 1 approvals obtained (waiting on @stephen-soltesz)

___
*[`deploy.sh` line 65 at r1](https://reviewable.io/reviews/m-lab/autonode/27#-OCkSCicB8iPgciWuAEp:-ODD6PzUEwUUTyBku_-W:be7epvn) ([raw file](https://github.com/m-lab/autonode/blob/9c0f986271577c5f1f43e707a107e524ac2f1cb9/deploy.sh#L65)):*
<details><summary><i>Previously, stephen-soltesz (Stephen Soltesz) wrote…</i></summary><blockquote>

It just seems strange for there to be a disconnect between uplink and interface maxrate.

I seem to recall at the very beginning of including the VMs I may have advocated for starting with 1g (including the max rate) so we wouldn't direct too much traffic there to start. We increased the probability and max rate for vms but never revisited this annotation.

IIRC, our VMs have a max egress rate of 7g? Is that correct? Could the "UPLINK" value match the max egress rates for our various configurations? https://cloud.google.com/compute/docs/network-bandwidth#vm-out 

I guess "UPLINK" is a remnant of switch deployments. Today this is something more generic like, "EGRESS" or MAXEGRESS or similar.
</blockquote></details>

Yeah, my understanding is just what you said, that GCE VMs have a max egress rate of 7Gbps. I'd be happy to update this annotation from 1g to 7g. It's true that "uplink" isn't perhaps the most expressive term we could use in a world with physical machines and VMs. I think it still mostly fits, however. I'm not opposed to changing the name either, but would like to maybe consider that separately. I update the value in deploy.sh to 7g.
___
*[`examples/ndt-fullstack.yml` line 29 at r1](https://reviewable.io/reviews/m-lab/autonode/27#-OCkSIHk7bkxuTdfUusC:-ODDASfvEw6yUgTI3EPd:bhfsgqn) ([raw file](https://github.com/m-lab/autonode/blob/9c0f986271577c5f1f43e707a107e524ac2f1cb9/examples/ndt-fullstack.yml#L29)):*
<details><summary><i>Previously, stephen-soltesz (Stephen Soltesz) wrote…</i></summary><blockquote>

I think the `stable` tag was a mistake. I would suggest we remove it and not include any remote update path to avoid the operational overhead of tracking a separate deployment path, and this would ensure that deployed configurations are stable. We would only need to maintain API compatibility in the Autojoin API.

I meant autonode because (in this model) M-Lab wants to be able to deploy a new version of the autonode to test it before promoting that to "stable" for all deployments. But, I think this is a bad idea. I think it would be preferable for there to just be the latest static version tag that autonode uses. Or, if deployed versions don't use the latest, it just continues to use the older version.
</blockquote></details>

I updated the env file to prompt users to manually edit DOCKER_TAG with the latest release tag of the autojoin repository.


<!-- Sent from Reviewable.io -->

stephen-soltesz reviewed (approved):
The other updates <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

I ultimately recommend eliminating the `DOCKER_TAG` variable.

> *[Reviewable](https://reviewable.io/reviews/m-lab/autonode/27#-:-ODMnIHi8fKPDHafI9ir:bwtku0g)* status: <span>:shipit:</span> complete! 1 of 1 approvals obtained (waiting on @nkinkade)

___
*[`deploy.sh` line 65 at r1](https://reviewable.io/reviews/m-lab/autonode/27#-OCkSCicB8iPgciWuAEp:-ODMm6Cr9mieYAjfSqCl:blsl5lh) ([raw file](https://github.com/m-lab/autonode/blob/9c0f986271577c5f1f43e707a107e524ac2f1cb9/deploy.sh#L65)):*
<details><summary><i>Previously, nkinkade wrote…</i></summary><blockquote>

Yeah, my understanding is just what you said, that GCE VMs have a max egress rate of 7Gbps. I'd be happy to update this annotation from 1g to 7g. It's true that "uplink" isn't perhaps the most expressive term we could use in a world with physical machines and VMs. I think it still mostly fits, however. I'm not opposed to changing the name either, but would like to maybe consider that separately. I update the value in deploy.sh to 7g.
</blockquote></details>

That all sounds good to me. I agree all of what I'm implying doesn't need to be in this PR.
___
*[`examples/ndt-fullstack.yml` line 29 at r1](https://reviewable.io/reviews/m-lab/autonode/27#-OCkSIHk7bkxuTdfUusC:-ODMmWuy9XAoqKwSaTor:b-8vk6n8) ([raw file](https://github.com/m-lab/autonode/blob/9c0f986271577c5f1f43e707a107e524ac2f1cb9/examples/ndt-fullstack.yml#L29)):*
<details><summary><i>Previously, nkinkade wrote…</i></summary><blockquote>

I updated the env file to prompt users to manually edit DOCKER_TAG with the latest release tag of the autojoin repository.
</blockquote></details>

I recommend not exposing `DOCKER_TAG` to users at all. I recommend the reference docker compose use a specific tagged version of register (which deploy.sh will override) like other containers. For now, a remote host updates a deployment by updating their version of the `ndt-fullstack.yml`, which would pull in the new version.


<!-- Sent from Reviewable.io -->

nkinkade committed:
Removes DOCKER_TAG variable from env file

Removes the DOCKER_TAG variable from the file env, and instead just statically
configures the latest released version of the autojoin repo (v0.2.11) in the
ndt-fullstack.yml Docker compose file.

nkinkade reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/m-lab/autonode/27)* status: <span>:shipit:</span> complete! 1 of 1 approvals obtained (waiting on @stephen-soltesz)

___
*[`examples/ndt-fullstack.yml` line 29 at r1](https://reviewable.io/reviews/m-lab/autonode/27#-OCkSIHk7bkxuTdfUusC:-ODMu0U2BNRNECu-lCC-:bcctn0m) ([raw file](https://github.com/m-lab/autonode/blob/9c0f986271577c5f1f43e707a107e524ac2f1cb9/examples/ndt-fullstack.yml#L29)):*
<details><summary><i>Previously, stephen-soltesz (Stephen Soltesz) wrote…</i></summary><blockquote>

I recommend not exposing `DOCKER_TAG` to users at all. I recommend the reference docker compose use a specific tagged version of register (which deploy.sh will override) like other containers. For now, a remote host updates a deployment by updating their version of the `ndt-fullstack.yml`, which would pull in the new version.
</blockquote></details>

I removed DOCKER_TAG from the file `env` and just statically configured "v0.2.11" in the ndt-fullstack.yml file.


<!-- Sent from Reviewable.io -->
